lib/conrad/collector.rb

```diff
@@ -0,0 +1,139 @@
+require 'conrad/formatters/json'
+require 'conrad/emitters/stdout'
+require 'conrad/processor_stack'
+
+module Conrad
+  # Used to collect a batch of events and send them using the configured
+  # emitter. This is especially useful for event recording done as part of a
+  # request/response cycle. Events can be collected here then sent at the end of
+  # the request cycle.
+  #
+  # Configuration is set via the class level and individual instances are
+  # spawned that use this configuration. The exception to that is the
+  # `event_metadata` which should be per collector using:
+  # `Conrad::Collector.current.event_metadata = { whatever: 'you want' }`
+  class Collector
+    # Used to setup metadata that will be added to every event in the collection
+    attr_accessor :event_metadata
+
+    # The events stored in the collector
+    attr_reader :events
+
+    class << self
+      # The set of Processors to use for each event added to the Collector
+      attr_writer :processors
+
+      # The Formatter to format each event added to the Collector
+      attr_writer :formatter
+
+      # The Emitter to use for sending events to another place. This should
+      # respond to #call. If `emit_as_batch` is set to true, this method *must*
+      # accept an Array of events. Otherwise, it is expected to accept a Hash.
+      attr_writer :emitter
```

Comment:
should these be class or instance level? Is there ever a case that the same process/codebase uses different emitters?
Like, what if there's a codepath that emits to sqs and another one that pushes to kafka? or is that contrived?

Reply:
Dunno! I think it's possible. It gets a bit trickier/grosser if that's the case if people have to manage it themselves, but we can. We could always provide the ability to override it if we wanted to.
```ruby
Conrad::Collector.emitter = Sqs.new

# some other code path
Conrad::Collector.current.emitter = Kafka.new
```

Or the architecture I've chosen is wrong, but it's what made sense to me if `Conrad::Collector.current` has to be in charge of initializing things. It also makes the `reset_state!` do a little more by reverting to the class configured ones for formatters, emitters, and processors, too

Reply:
I don't think your architecture is wrong at all. Can we delegate these to `current` as to avoid potential conflicts across threads? Especially if consumers decide to have formatters/emitters to konw things about the events they process: instance variables & such. Delegation to the `current` instance would ensure thread safety, right?

Reply:
Hm. I'm not sure I understand what you mean by delegation to the Current. At the moment what happens is this:
1. Someone calls `Conrad::Collector.current` to get an instance of collector for their thread
2. When using it, it by default uses the processor, formatter, and emitter defined on the class.
3. If we allowed overriding, it would override for that particular, current instance (e.g., `Conrad::Collector.current.emitter = Kafka.new`). This means that for that thread, for the lifespan before emitting events, the collector would be configured as such. 

For thread safety, we could initialize to the class variables _as configured_ at the time of initialization. If they change underneath the collector it won't matter; it will be on the original configuration. The question is: do we still reset to the class configured values in `reset_state!`?

Reply:
What if a company wants to send all of their audit logs to one place and their financial purchase logs somewhere else? I like Greg's point about making them instance vars - whatcha hoping to benefit from by using class vars instead?

Reply:
I think these concerns are alleviated by the initializer. It defaults to using the class configured values but allows passing in different values to the initializer. Adding writer for those values would not be difficult.

Reply:
Won't they clobber one another if you have 2 threads that need different emitters but write to the same class variable? And since the initializer is `||=` the emitters won't get reset, right? Or am I missing something?

Reply:
We talked in person, but the resolution is to prefix the class instance variables with `default_` so that the intention is clearer.